client/core/trade.go

```diff
@@ -1239,14 +1244,111 @@ func (t *trackedTrade) isSwappable(ctx context.Context, match *matchTracker) (re
 				"but we have not seen the counter-party's redemption yet. This could just"+
 				" be network latency.", match)
 		}
+
 		match.setSwapConfirms(int64(confs))
+
 		t.notify(newMatchNote(TopicConfirms, "", "", db.Data, t, match))
 		return false, false
 	}
 
 	return false, false
 }
 
+// updateFeesPaid starts a goroutine that updates the fees used for dynamic fee
+// checker transactions. We do not know the exact fees the tx will use until
+// they are mined, so this waits until they are mined and updates the value for
+// the entire trade.
+//
+// TODO: This method does not persist if the client is shut down. It probably
+// should because payed fees will be incorrect forever for this trade if not.
+//
+// This method accesses match fields and MUST be called with the trackedTrade
+// mutex lock held for reads.
+func (t *trackedTrade) updateFeesPaid(ctx context.Context, match *matchTracker, isInit bool) {
+	wallet := t.wallets.fromWallet
+	if !isInit {
+		wallet = t.wallets.toWallet
+	}
+	feeChecker, dynamic := wallet.Wallet.(asset.DynamicTxFeeChecker)
+	if !dynamic {
+		return
+	}
+	go func() {
+		ctx, cancel := context.WithTimeout(ctx, updateFeesTimeout)
+		defer cancel()
+		txType := "swap"
+		if !isInit {
+			txType = "redemption"
+		}
+		var coinID []byte
+
+		// Check if a swap or redeem coin id has been populated in the
+		// match tracker. If it has we ask the wallet for the fees paid
+		// and add that to either the total swap or redeem fees for the
+		// trade.
+		check := func() (doReturn bool) {
+			t.mtx.RLock()
+			if isInit {
+				coinID = []byte(match.MetaData.Proof.MakerSwap)
+				if match.Side != order.Maker {
+					coinID = []byte(match.MetaData.Proof.TakerSwap)
+				}
+			} else {
+				coinID = []byte(match.MetaData.Proof.MakerRedeem)
+				if match.Side != order.Maker {
+					coinID = []byte(match.MetaData.Proof.TakerRedeem)
+				}
+			}
+			t.mtx.RUnlock()
+			if coinID == nil {
+				return false
+			}
+			actualSwapFees, err := feeChecker.TransactionFeesPaid(ctx, coinID)
+			if err != nil {
+				if errors.Is(err, asset.CoinNotFoundError) {
+					return false
+				}
+				t.dc.log.Errorf("Failed to determine actual %s transaction fees paid for "+
+					"match %s: %v", txType, match, err)
+				return true
+			}
+			t.mtx.Lock()
+			if isInit {
+				t.metaData.SwapFeesPaid += actualSwapFees
```

Comment:
Didn't we already `+=` in `swapMatchGroup`. We're not subtracting that original value, are we?

I think we might need to move the `SwapFeesPaid` and `RedemptionFeesPaid` fields to the `MatchMetadata` if we're going to update the fees?

Reply:
There's a `!dynamic` guarding the `+=` in `swapMatchGroup`